core/node/api_server/src/execution_sandbox/execute.rs

```diff
@@ -114,72 +51,99 @@ impl TransactionExecutor {
         let (env, storage) =
             apply::prepare_env_and_storage(connection, setup_args, &block_args).await?;
         let state_override = state_override.unwrap_or_default();
         let storage = StorageWithOverrides::new(storage, &state_override);
 
-        let (published_bytecodes, execution_result) = self
-            .inspect_transaction_with_bytecode_compression(storage, env, execution_args, tracers)
+        let result = self
+            .inspect_transaction_with_bytecode_compression(
+                storage,
+                env,
+                execution_args,
+                tracing_params,
+            )
             .await?;
         drop(vm_permit);
 
         let metrics =
-            vm_metrics::collect_tx_execution_metrics(total_factory_deps, &execution_result);
+            vm_metrics::collect_tx_execution_metrics(total_factory_deps, &result.tx_result);
         Ok(TransactionExecutionOutput {
-            vm: execution_result,
+            vm: *result.tx_result,
+            call_traces: result.call_traces,
             metrics,
-            are_published_bytecodes_ok: published_bytecodes.is_ok(),
+            are_published_bytecodes_ok: result.compression_result.is_ok(),
         })
     }
 }
 
+impl From<MockOneshotExecutor> for TransactionExecutor {
+    fn from(executor: MockOneshotExecutor) -> Self {
+        Self::Mock(executor)
+    }
+}
+
 #[async_trait]
 impl<S> OneshotExecutor<S> for TransactionExecutor
 where
     S: ReadStorage + Send + 'static,
 {
-    type Tracers = Vec<ApiTracer>;
-
-    async fn inspect_transaction(
+    async fn inspect_transaction_with_bytecode_compression(
         &self,
         storage: S,
         env: OneshotEnv,
         args: TxExecutionArgs,
-        tracers: Self::Tracers,
-    ) -> anyhow::Result<VmExecutionResultAndLogs> {
+        tracing_params: OneshotTracingParams,
+    ) -> anyhow::Result<OneshotTransactionExecutionResult> {
         match self {
             Self::Real(executor) => {
                 executor
-                    .inspect_transaction(storage, env, args, tracers)
+                    .inspect_transaction_with_bytecode_compression(
+                        storage,
+                        env,
+                        args,
+                        tracing_params,
+                    )
+                    .await
+            }
+            Self::Mock(executor) => {
+                executor
+                    .inspect_transaction_with_bytecode_compression(
+                        storage,
+                        env,
+                        args,
+                        tracing_params,
+                    )
                     .await
             }
-            Self::Mock(executor) => executor.inspect_transaction(storage, env, args, ()).await,
         }
     }
+}
 
-    async fn inspect_transaction_with_bytecode_compression(
+#[async_trait]
+impl<S> TransactionValidator<S> for TransactionExecutor
```

Comment:
It seems possible to use dynamically dispatched executor instead (`Box<dyn TransactionValidator< StorageWithOverrides>>`), but this can have performance implications, so haven't done that.